src/UIComponents.cpp

```diff
@@ -0,0 +1,91 @@
+#include "UIComponents.hpp"
+
+UIEntity::~UIEntity()
+{
+  entity->detach((IObserver*)this);
+};
+
+UIEntity::UIEntity(std::shared_ptr<HAEntity> _entity, lv_obj_t* _parent)
+{
+  parentContainer = _parent;
+  entity = _entity;
+  entity->attach((IObserver*)this);
+};
+
+UIButton::UIButton(std::shared_ptr<HAEntity> _entity, lv_obj_t* _parent) :
+  UIEntity(_entity, _parent)
+{
+  // START BUTTON EXAMPLE
+  btn = lv_btn_create(_parent); /*Add a button the current screen*/
+  lv_obj_set_size(btn, 240, 50);
+  lv_obj_center(btn);
+  lv_obj_add_flag(btn, LV_OBJ_FLAG_CHECKABLE);
+  lv_obj_add_event_cb(btn, UIButton::btn_press_cb, LV_EVENT_ALL, reinterpret_cast<void*>(&entity)); /*Assign a callback to the button*/
+
+  lv_obj_t* label = lv_label_create(btn); /*Add a label to the button*/
+  lv_label_set_text(label, _entity->name.c_str()); /*Set the labels text*/
+  lv_obj_center(label);
+  uiupdate();
+};
+
+void UIButton::uiupdate()
+{
+  auto state = entity->getJsonState();
+  if (state["state"] == "on") {
+    lv_obj_add_state(btn, LV_STATE_CHECKED);
+  }
+  else {
+    lv_obj_clear_state(btn, LV_STATE_CHECKED);
+  }
```

Comment:
similar to things you mentioned on IRC, this set of lines is a fine candidate for some kind of reuse

Reply:
Added a fixme to this, because yes - we should replace this code later.